zerver/forms.py

```diff
@@ -279,7 +279,10 @@ def clean(self) -> Dict[str, Any]:
 
         if username is not None and password:
             subdomain = get_subdomain(self.request)
-            realm = get_realm(subdomain)
+            try:
+                realm = get_realm(subdomain)  # type: Optional[Realm]
+            except Realm.DoesNotExist:
+                realm = None
```

Comment:
It's possible we want to raise an exception here; worth investigating.

Reply:
Raising an exception here breaks `test_login_invalid_subdomain`.

I agree that something’s weird about this, though. Why should each individual authentication backend be responsible for checking that the realm is valid? There’s an odd [comment](https://github.com/zulip/zulip/blob/2.0.3/zproject/backends.py#L128) in `ZulipDummyBackend`:

```python
    def authenticate(self, username: Optional[str]=None, realm: Optional[Realm]=None,
                     use_dummy_backend: bool=False,
                     return_data: Optional[Dict[str, Any]]=None) -> Optional[UserProfile]:
        if use_dummy_backend:
            # These are kwargs only for readability; they should never be None
            assert username is not None
            assert realm is not None
            return common_get_active_user(username, realm, return_data)
        return None
```

which suggests a mistaken assumption that kwargs need to have default values…is that why?

Reply:
I refactored things so `authenticate` always takes `realm: Realm` instead of `realm: Optional[Realm]=None` and the caller is responsible for the check.

Reply:
So, Django `authenticate` methods are super evil in that only the ones for registered backends who function signature matches the call site get called.  So this is likely in part encoding the assumption that `use_dummy_backend=True` => `username is not None`.  

That said, this code was definitely written before the Python 3 migration, and we haven't attempted to do an audit for kwargs without default values since then.

Reply:
(And third-party authentication backends like `python-social-auth` have their own call signatures they use when calling these).  So we may need to separate mandatory kwargs here.

Reply:
I saw the Django signature checking code, but I don’t see that as an explanation. The default value of `username` had the effect of allowing Django to call this function if a `username` wasn’t passed. But what is the point of allowing that?

I also don’t see how Python 3 is relevant. In both Python 2 and Python 3, a parameter declared with or without a default value can be passed as either a positional argument or a keyword argument. (Python 3 adds the ability to explicitly declare keyword-only arguments, but it’s not in use here.)

Reply:
I am not super comfortable with just making these positional arguments passed by keyword -- that creates a serious risk that someone making changes to the code in the future thinks it's reasonable to pass arguments to `authenticate` positionally, and e.g. accidentally calls `authenticate(foo, bar, baz)` in a way that accidentally makes every login succeed.  

More generally, we've avoided use of the "passing an argument declared positionally by kwarg name" pattern in Python because many less experienced Python programmers don't understand it, and we want our code to be readable by a wide range of programmers.

The reason I mentioned Python 3 is a potentially cleaner way to achieve our goals for these parameters would be to make these be keyword-only arguments, which is what I thought you might be proposing?  

The mypy type-signature checking on these methods is basically useless, because they're not called directly anyway.  

In any case, probably this commit or any other auth backend logic refactoring should be serialized after https://github.com/zulip/zulip/pull/11450/files, since it'll merge conflict heavily with it and #11450 deletes a bunch of the code this modifies.

Reply:
So, I think you’re just confirming my theory. In the interest of being totally clear:

There is no such thing as “positional arguments passed by keyword”, and no such thing as “declared positionally”.

It is _already possible_ to call `def authenticate(username=None)` as `authenticate(foo)`, and not only is it already possible, we’re [already doing it](https://github.com/zulip/zulip/blob/2.0.3/zerver/tests/test_auth_backends.py#L2340) in some cases.

Declaring the function with a default value does nothing to prevent that, nor does your imagined stylistic convention if it’s not enforced.

Meanwhile, the declared default value does have the real downside of communicating that it’s okay to omit the parameter, and then forcing us to handle (and test for coverage) extra conditions that should never have existed, and then allowing other code to grow reliant upon those extra conditions.

We can make these keyword-only arguments if you’d like, but it wasn’t something I proposed.

Reply:
Added a commit to make all `authenticate` parameters keyword-only.

Reply:
#11450 looks stalled. Would you like me to split out the google_oauth2 related parts of this into separate commits?